R/openapi-types.R

```diff
@@ -2,114 +2,197 @@
+      # > value quotes/h2g2.txt will be sent as quotes%2Fh2g2.txt
+      openApiRegex = openApiRegex,
+      parser = parser,
+
+      # TODO - not keen on these array functions
+      # believe they are used only for comma separated arrays inside dynamic routes
+      # - not sure Plumber needs to support these
+      openApiRegexArray = paste0("(?:(?:", openApiRegex, "),?)+"),
+      # TODO - this won't work for strings that contain commas?
+      parserArray = function(x) {parser(stri_split_fixed(x, ",")[[1]])}
+    )
+
+    for (apiType in keys) {
+      apiTypesInfo[[apiType]] <<- entry
     }
-    # make sure it could be called again
-    plumberToApiTypeMap[[apiType]] <<- apiType
 
     invisible(TRUE)
   }
 
   addApiInfo(
-    "boolean",
-    c("bool", "boolean", "logical"),
-    "[01tfTF]|true|false|TRUE|FALSE",
-    as.logical,
+    c("boolean", "bool", "logical"),
+    openApiType = "boolean",
+    openApiRegex = "[01tfTF]|true|false|TRUE|FALSE",
+    parser = as.logical,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("number", "numeric"),
+    openApiType = "number",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("double", "dbl"),
+    openApiType = "number",
+    openApiFormat = "double",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("float"),
+    openApiType = "number",
+    openApiFormat = "float",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("integer", "int"),
+    openApiType = "integer",
+    openApiFormat = "int64",
+    openApiRegex = "-?\\\\d+",
+    parser = as.integer,
     location = c("query", "path")
   )
+
   addApiInfo(
-    "number",
-    c("dbl", "double", "float", "number", "numeric"),
-    "-?\\\\d*\\\\.?\\\\d+",
-    as.numeric,
-    format = "double",
+    c("string", "str", "chr", "character"),
+    openApiType = "string",
+    openApiRegex = "[^/]+",
+    parser = as.character,
     location = c("query", "path")
   )
+
+  # be careful not to trap lubridate on package installation
+  lubridate_available <- function() {
+    system.file(package = "lubridate") != "";
+  }
+
+  parse_datetime <- function(x) {
+    if (lubridate_available()) lubridate::as_datetime(x) else as.Date.POSIXct(x)
+  }
+
+  parse_date <- function(x) {
+    if (lubridate_available()) lubridate::as_date(x) else as.Date(x)
+  }
+
   addApiInfo(
-    "integer",
-    c("int", "integer"),
-    "-?\\\\d+",
-    as.integer,
-    format = "int64",
+    c("date-time", "datetime"),
+    openApiType = "string",
+    # much more complex regexes are available... e.g. see https://stackoverflow.com/a/3143231/373321
+    # (if importing one of those, remember to use cc attribution)
+    openApiRegex = "\\\\d{4}-\\\\d{2}-\\\\d{2}T\\\\d{2}:\\\\d{2}:\\\\d{2}Z",
```

Comment:
Is this intentionally limiting the timezone to `Z`?

Reply:
I'm very happy to see it restricted... I encourage everyone to use UTC whenever they can (and if I can persuade the UK government to drop BST then it'd make my life so much easier...)

But from a [RFC 3339](https://www.rfc-editor.org/rfc/rfc3339#section-5.6) perspective, it could probably be extended.

As far as I understand it, this regex is only used in parsing path query variables (but I might have that wrong...). With that in mind, are these regexes needed at all?

If they are, then the comment points to [stackoverflow.com/a/3143231/373321](https://stackoverflow.com/a/3143231/373321) with a wealth of options :)